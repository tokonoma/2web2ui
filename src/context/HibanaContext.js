import React, { createContext, useContext } from 'react';
import { connect } from 'react-redux';
import { updateUserUIOptions } from 'src/actions/currentUser';
import { showAlert } from 'src/actions/globalAlert';
import { isUserUiOptionSet } from 'src/helpers/conditions/user';
import { isAccountUiOptionSet } from 'src/helpers/conditions/account';
import { selectCondition } from 'src/selectors/accessConditionState';

const HibanaStateContext = createContext();

function Provider(props) {
  const { isHibanaEnabled: _isHibanaEnabled, children, ...rest } = props;

  // TODO: When OG theme is removed, the env variables will no longer be necessary
  const isHibanaEnabled =
    process.env.REACT_APP_DEFAULT_TO_HIBANA === 'true' ? true : _isHibanaEnabled;

  return (
    <HibanaStateContext.Provider value={{ isHibanaEnabled, ...rest }}>
      {children}
    </HibanaStateContext.Provider>
  );
}

function mapStateToProps(state) {
  return {
    isHibanaEnabled: selectCondition(isUserUiOptionSet('isHibanaEnabled'))(state),
    isBannerVisible: selectCondition(isUserUiOptionSet('isHibanaBannerVisible'))(state),
    hasThemeControls: selectCondition(isAccountUiOptionSet('hibana.hasThemeControls'))(state),
  };
}

const mapDispatchToProps = {
  dismissBanner: () => updateUserUIOptions({ isHibanaBannerVisible: false }),
  setIsHibanaEnabled: bool => {
    if (window.pendo && window.pendo.track) {
      window.pendo.track(`Hibana Toggle - ${Boolean(bool) ? 'On' : 'Off'}`);
    }

    // Always dismiss the banner when the user toggles their theme
    return updateUserUIOptions({ isHibanaEnabled: bool, isHibanaBannerVisible: false });
  },
  showAlert,
};

export function useHibana() {
  const context = useContext(HibanaStateContext);

  if (context === undefined) throw new Error('useHibana must be used within a HibanaProvider');

  return [context];
}

export const HibanaProvider = connect(mapStateToProps, mapDispatchToProps)(Provider);

export function HibanaConsumer({ children }) {
  return <HibanaStateContext.Consumer>{children}</HibanaStateContext.Consumer>;
}
