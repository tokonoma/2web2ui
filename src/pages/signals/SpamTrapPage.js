/* eslint-disable max-lines */
import React, { useState } from 'react';
import { getSpamHits } from 'src/actions/signals';
import { selectSpamHitsDetails } from 'src/selectors/signals';
import { PageLink } from 'src/components/links';
import { Box, Grid, Panel, Stack } from 'src/components/matchbox';
import { useHibana } from 'src/context/HibanaContext';
import Page from './components/SignalsPage';
import BarChart from './components/charts/barchart/BarChart';
import SpamTrapActions from './components/actionContent/SpamTrapActions';
import TooltipMetric from 'src/components/charts/TooltipMetric';
import DateFilter from './components/filters/DateFilter';
import { SPAM_TRAP_INFO } from './constants/info';
import withDetails from './containers/withDetails';
import withDateSelection from './containers/withDateSelection';
import { Loading } from 'src/components';
import Callout from 'src/components/callout';
import Legend from './components/charts/legend/Legend';
import Calculation from './components/viewControls/Calculation';
import ChartHeader from './components/ChartHeader';
import { formatFullNumber, formatNumber, roundToPlaces } from 'src/helpers/units';
import moment from 'moment';
import _ from 'lodash';
import { spamTrapHitTypesCollection, spamTrapHitTypesByLabel } from './constants/spamTrapHitTypes';

export function SpamTrapPage(props) {
  const [state] = useHibana();
  const { isHibanaEnabled } = state;
  const [calculation, setCalculation] = useState('relative');

  const handleCalculationToggle = value => {
    setCalculation(value);
  };

  const getYAxisProps = () => {
    const { data } = props;

    return {
      tickFormatter:
        calculation === 'relative'
          ? tick => `${roundToPlaces(tick * 100, 3)}%`
          : tick => formatNumber(tick),
      domain:
        data.every(({ relative_trap_hits }) => !relative_trap_hits) && calculation === 'relative'
          ? [0, 1]
          : ['auto', 'auto'],
    };
  };

  const getXAxisProps = () => {
    const { xTicks } = props;

    return {
      ticks: xTicks,
      tickFormatter: tick => moment(tick).format('M/D'),
    };
  };

  const getTooltipContent = ({ payload = {} }) => (
    <Stack>
      {calculation === 'absolute' ? (
        <TooltipMetric label="Spam Trap Hits" value={formatFullNumber(payload.trap_hits)} />
      ) : (
        <TooltipMetric
          label="Spam Trap Rate"
          value={`${roundToPlaces(payload.relative_trap_hits * 100, 4)}%`}
        />
      )}

      <Stack>
        {spamTrapHitTypesCollection.map(({ OGFill, hibanaFill, key, label }) => {
          const fill = isHibanaEnabled ? hibanaFill : OGFill;

          return (
            <TooltipMetric
              color={fill}
              key={key}
              label={label}
              value={
                calculation === 'absolute'
                  ? `${formatFullNumber(payload[key])}`
                  : `${roundToPlaces(payload[`relative_${key}`] * 100, 4)}%`
              }
            />
          );
        })}
      </Stack>
      <TooltipMetric label="Injections" value={formatFullNumber(payload.injections)} />
    </Stack>
  );

  const renderContent = () => {
    const {
      data = [],
      handleDateSelect,
      loading,
      gap,
      empty,
      error,
      selectedDate,
      handleDateHover,
      resetDateHover,
      hoveredDate,
      shouldHighlightSelected,
    } = props;
    const selectedSpamTrapHits = _.find(data, ['date', selectedDate]) || {};
    let chartPanel;

    if (empty) {
      chartPanel = (
        <Callout title="No Data Available">Insufficient data to populate this chart</Callout>
      );
    }

    if (error) {
      chartPanel = <Callout title="Unable to Load Data">{error.message}</Callout>;
    }

    if (loading) {
      chartPanel = (
        <div style={{ height: '220px', position: 'relative' }}>
          <Loading />
        </div>
      );
    }

    return (
      <Grid>
        <Grid.Column sm={12} md={7}>
          <Panel sectioned>
            <ChartHeader
              title="Spam Trap Monitoring"
              primaryArea={
                <Calculation initialSelected={calculation} onChange={handleCalculationToggle} />
              }
              tooltipContent={SPAM_TRAP_INFO}
            />
            {chartPanel || (
              <div className="LiftTooltip">
                <BarChart
                  gap={gap}
                  onClick={handleDateSelect}
                  selected={selectedDate}
                  timeSeries={data}
                  onMouseOver={handleDateHover}
                  onMouseOut={resetDateHover}
                  hovered={hoveredDate}
                  shouldHighlightSelected={shouldHighlightSelected}
                  tooltipContent={getTooltipContent}
                  yKeys={spamTrapHitTypesCollection
                    .map(({ key, OGFill, hibanaFill }) => {
                      const fill = isHibanaEnabled ? hibanaFill : OGFill;

                      return {
                        fill,
                        key: calculation === 'relative' ? `relative_${key}` : key,
                      };
                    })
                    .reverse()}
                  yAxisProps={getYAxisProps()}
                  xAxisProps={getXAxisProps()}
                />
                <Legend
                  items={spamTrapHitTypesCollection}
                  tooltipContent={label => spamTrapHitTypesByLabel[label].description}
                />
              </div>
            )}
          </Panel>
        </Grid.Column>
        <Grid.Column sm={12} md={5} mdOffset={0}>
          {!loading && (
            <Box as={Panel} sectioned>
              <div>
                {!chartPanel && (
                  <SpamTrapActions
                    percent={selectedSpamTrapHits.relative_trap_hits}
                    date={selectedDate}
                  />
                )}
              </div>
            </Box>
          )}
        </Grid.Column>
      </Grid>
    );
  };

  const { facet, facetId, subaccountId } = props;

  return (
    <Page
      breadcrumbAction={{
        content: 'Back to Spam Trap Overview',
        to: '/signals/spam-traps',
        component: PageLink,
      }}
      title="Spam Traps"
      facet={facet}
      facetId={facetId}
      subaccountId={subaccountId}
    >
      <Panel title="Spam Trap Monitoring">
        <Panel.Section>
          <Grid>
            <Grid.Column xs={12} md={4}>
              <DateFilter label="Date Range" />
            </Grid.Column>
          </Grid>
        </Panel.Section>
      </Panel>
      {renderContent()}
    </Page>
  );
}

export default withDetails(withDateSelection(SpamTrapPage), { getSpamHits }, selectSpamHitsDetails);
