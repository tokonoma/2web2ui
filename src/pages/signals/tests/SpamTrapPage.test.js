import React from 'react';
import { shallow } from 'enzyme';
import { useHibana } from 'src/context/HibanaContext';
import { SpamTrapPage } from '../SpamTrapPage';

jest.mock('src/context/HibanaContext');

useHibana.mockImplementation(() => [{ isHibanaEnabled: true }]);

describe('Signals Spam Trap Page', () => {
  const data = [
    {
      date: '2017-01-01',
      injections: 1240,
      relative_trap_hits: 0.1,
      relative_trap_hits_parked: 0.025,
      relative_trap_hits_recycled: 0.025,
      relative_trap_hits_typo: 0.05,
      trap_hits: 124,
      trap_hits_parked: 31,
      trap_hits_recycled: 31,
      trap_hits_typo: 62,
    },
    {
      date: '2017-01-02',
      injections: 424,
      relative_trap_hits: 0.5,
      relative_trap_hits_parked: 0.125,
      relative_trap_hits_recycled: 0.125,
      relative_trap_hits_typo: 0.25,
      trap_hits: 212,
      trap_hits_parked: 53,
      trap_hits_recycled: 53,
      trap_hits_typo: 106,
    },
  ];

  const defaultProps = {
    facetId: 'test.com',
    facet: 'sending-domain',
    data,
    gap: 0.25,
    handleDateSelect: jest.fn(),
    loading: false,
    empty: false,
    selectedDate: '2017-01-01',
    xTicks: [],
  };

  const subject = props => {
    return shallow(<SpamTrapPage {...defaultProps} {...props} />);
  };

  it('renders correctly', () => {
    const wrapper = subject();

    expect(wrapper).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    const wrapper = subject({ loading: true });

    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('renders empty correctly', () => {
    const wrapper = subject({ empty: true });

    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('renders error correctly', () => {
    const wrapper = subject({ error: { message: 'error message' } });

    expect(wrapper).toMatchSnapshot();
  });

  describe('local state', () => {
    it('handles calculation type', () => {
      const wrapper = subject();
      const calculation = shallow(wrapper.find('ChartHeader').props().primaryArea);

      calculation.simulate('change', 'absolute');

      expect(wrapper.find('BarChart').prop('yKeys')).toEqual([
        expect.objectContaining({ key: 'trap_hits_typo' }),
        expect.objectContaining({ key: 'trap_hits_recycled' }),
        expect.objectContaining({ key: 'trap_hits_parked' }),
      ]);
    });
  });

  describe('bar chart props', () => {
    it('renders tooltip content', () => {
      const wrapper = subject();
      const Tooltip = wrapper.find('BarChart').prop('tooltipContent');

      expect(shallow(<Tooltip payload={data[0]} />)).toMatchSnapshot();
    });

    it('gets x axis props', () => {
      const wrapper = subject({ calculation: 'relative' });
      const calculation = shallow(wrapper.find('ChartHeader').props().primaryArea);
      calculation.simulate('change', 'absolute');
      const axisProps = wrapper.find('BarChart').prop('xAxisProps');
      expect(axisProps).toMatchSnapshot();
      expect(axisProps.tickFormatter('2018-12-05')).toEqual('12/5');
    });

    it('gets y axis props with relative calculation', () => {
      const wrapper = subject({ calculation: 'absolute' });
      const calculation = shallow(wrapper.find('ChartHeader').props().primaryArea);
      calculation.simulate('change', 'relative');
      const axisProps = wrapper.find('BarChart').prop('yAxisProps');
      expect(axisProps.tickFormatter(0.2468)).toEqual('24.68%');
    });

    it('gets y axis props with absolute calculation', () => {
      const wrapper = subject({ calculation: 'relative' });
      const calculation = shallow(wrapper.find('ChartHeader').props().primaryArea);

      calculation.simulate('change', 'absolute');

      const axisProps = wrapper.find('BarChart').prop('yAxisProps');

      expect(axisProps.tickFormatter(2468)).toEqual('2.47K');
      expect(axisProps.domain).toEqual(['auto', 'auto']);
    });

    it('gets y axis props with relative calculation and 0 values', () => {
      const wrapper = subject({
        calculation: 'absolute',
        data: [{ relative_trap_hits: 0 }, { relative_trap_hits: null }],
      });
      const calculation = shallow(wrapper.find('ChartHeader').props().primaryArea);

      calculation.simulate('change', 'relative');

      const axisProps = wrapper.find('BarChart').prop('yAxisProps');

      expect(axisProps.domain).toEqual([0, 1]);
    });
  });
});
