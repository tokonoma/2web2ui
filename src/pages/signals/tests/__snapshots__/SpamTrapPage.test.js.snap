// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals Spam Trap Page bar chart props gets x axis props 1`] = `
Object {
  "tickFormatter": [Function],
  "ticks": Array [],
}
`;

exports[`Signals Spam Trap Page bar chart props renders tooltip content 1`] = `
<Stack>
  <TooltipMetric
    label="Spam Trap Rate"
    value="10%"
  />
  <Stack>
    <TooltipMetric
      color="#1a497d"
      key="trap_hits_parked"
      label="Parked"
      value="2.5%"
    />
    <TooltipMetric
      color="#1273e6"
      key="trap_hits_recycled"
      label="Recycled"
      value="2.5%"
    />
    <TooltipMetric
      color="#78b6ff"
      key="trap_hits_typo"
      label="Typo"
      value="5%"
    />
  </Stack>
  <TooltipMetric
    label="Injections"
    value="1,240"
  />
</Stack>
`;

exports[`Signals Spam Trap Page renders correctly 1`] = `
<SignalsPage
  breadcrumbAction={
    Object {
      "component": [Function],
      "content": "Back to Spam Trap Overview",
      "to": "/signals/spam-traps",
    }
  }
  facet="sending-domain"
  facetId="test.com"
  title="Spam Traps"
>
  <Panel
    title="Spam Trap Monitoring"
  >
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>
  <Grid>
    <Grid.Column
      md={7}
      sm={12}
    >
      <Panel
        sectioned={true}
      >
        <ChartHeader
          primaryArea={
            <Calculation
              initialSelected="relative"
              onChange={[Function]}
            />
          }
          title="Spam Trap Monitoring"
          tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
        />
        <div
          className="LiftTooltip"
        >
          <BarChart
            gap={0.25}
            height={250}
            margin={
              Object {
                "bottom": 5,
                "left": 18,
                "right": 0,
                "top": 12,
              }
            }
            onClick={[MockFunction]}
            selected="2017-01-01"
            shouldHighlightSelected={true}
            timeSeries={
              Array [
                Object {
                  "date": "2017-01-01",
                  "injections": 1240,
                  "relative_trap_hits": 0.1,
                  "relative_trap_hits_parked": 0.025,
                  "relative_trap_hits_recycled": 0.025,
                  "relative_trap_hits_typo": 0.05,
                  "trap_hits": 124,
                  "trap_hits_parked": 31,
                  "trap_hits_recycled": 31,
                  "trap_hits_typo": 62,
                },
                Object {
                  "date": "2017-01-02",
                  "injections": 424,
                  "relative_trap_hits": 0.5,
                  "relative_trap_hits_parked": 0.125,
                  "relative_trap_hits_recycled": 0.125,
                  "relative_trap_hits_typo": 0.25,
                  "trap_hits": 212,
                  "trap_hits_parked": 53,
                  "trap_hits_recycled": 53,
                  "trap_hits_typo": 106,
                },
              ]
            }
            tooltipContent={[Function]}
            width="99%"
            xAxisProps={
              Object {
                "tickFormatter": [Function],
                "ticks": Array [],
              }
            }
            xAxisRefLines={Array []}
            xKey="date"
            yAxisProps={
              Object {
                "domain": Array [
                  "auto",
                  "auto",
                ],
                "tickFormatter": [Function],
              }
            }
            yAxisRefLines={Array []}
            yKey="value"
            yKeys={
              Array [
                Object {
                  "fill": "#78b6ff",
                  "key": "relative_trap_hits_typo",
                },
                Object {
                  "fill": "#1273e6",
                  "key": "relative_trap_hits_recycled",
                },
                Object {
                  "fill": "#1a497d",
                  "key": "relative_trap_hits_parked",
                },
              ]
            }
            yRange={
              Array [
                "auto",
                "auto",
              ]
            }
          />
          <Legend
            items={
              Array [
                Object {
                  "OGFill": "#50CFDA",
                  "description": "
      This is an email address from a domain hosted by a parked domain provider. These addresses
      can point to list quality issues but do not affect your reputation.
    ",
                  "hibanaFill": "#1a497d",
                  "key": "trap_hits_parked",
                  "label": "Parked",
                },
                Object {
                  "OGFill": "#29B9C7",
                  "description": "
      An email address that once was a real address, but now is a spam trap. These addresses signal
      that you should clean out old unengaged recipients.
    ",
                  "hibanaFill": "#1273e6",
                  "key": "trap_hits_recycled",
                  "label": "Recycled",
                },
                Object {
                  "OGFill": "#1A838B",
                  "description": "
      This is an email address from a domain that looks like a real mailbox provider, like
      gmal.com. These addresses signal that you should improve your list acquisition practices.
    ",
                  "hibanaFill": "#78b6ff",
                  "key": "trap_hits_typo",
                  "label": "Typo",
                },
              ]
            }
            tooltipContent={[Function]}
          />
        </div>
      </Panel>
    </Grid.Column>
    <Grid.Column
      md={5}
      mdOffset={0}
      sm={12}
    >
      <Box
        as={[Function]}
        sectioned={true}
      >
        <div>
          <SpamTrapActions
            date="2017-01-01"
            percent={0.1}
          />
        </div>
      </Box>
    </Grid.Column>
  </Grid>
</SignalsPage>
`;

exports[`Signals Spam Trap Page renders empty correctly 1`] = `
Array [
  <Panel
    title="Spam Trap Monitoring"
  >
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>,
  <Panel
    sectioned={true}
  >
    <ChartHeader
      primaryArea={
        <Calculation
          initialSelected="relative"
          onChange={[Function]}
        />
      }
      title="Spam Trap Monitoring"
      tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
    />
    <Callout
      title="No Data Available"
    >
      Insufficient data to populate this chart
    </Callout>
  </Panel>,
]
`;

exports[`Signals Spam Trap Page renders error correctly 1`] = `
<SignalsPage
  breadcrumbAction={
    Object {
      "component": [Function],
      "content": "Back to Spam Trap Overview",
      "to": "/signals/spam-traps",
    }
  }
  facet="sending-domain"
  facetId="test.com"
  title="Spam Traps"
>
  <Panel
    title="Spam Trap Monitoring"
  >
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>
  <Grid>
    <Grid.Column
      md={7}
      sm={12}
    >
      <Panel
        sectioned={true}
      >
        <ChartHeader
          primaryArea={
            <Calculation
              initialSelected="relative"
              onChange={[Function]}
            />
          }
          title="Spam Trap Monitoring"
          tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
        />
        <Callout
          title="Unable to Load Data"
        >
          error message
        </Callout>
      </Panel>
    </Grid.Column>
    <Grid.Column
      md={5}
      mdOffset={0}
      sm={12}
    >
      <Box
        as={[Function]}
        sectioned={true}
      >
        <div />
      </Box>
    </Grid.Column>
  </Grid>
</SignalsPage>
`;

exports[`Signals Spam Trap Page renders loading correctly 1`] = `
Array [
  <Panel
    title="Spam Trap Monitoring"
  >
    <Panel.Section>
      <Grid>
        <Grid.Column
          md={4}
          xs={12}
        >
          <Connect(DateFilter)
            label="Date Range"
          />
        </Grid.Column>
      </Grid>
    </Panel.Section>
  </Panel>,
  <Panel
    sectioned={true}
  >
    <ChartHeader
      primaryArea={
        <Calculation
          initialSelected="relative"
          onChange={[Function]}
        />
      }
      title="Spam Trap Monitoring"
      tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
    />
    <div
      style={
        Object {
          "height": "220px",
          "position": "relative",
        }
      }
    >
      <Loading />
    </div>
  </Panel>,
]
`;
